File: tests/fixtures/graph-src/python/shapes.py
"""Shape helpers for the Python portion of the fixture.

Exercises: classes with methods and members, an enum, a module-level function,
and a type alias.
"""

from __future__ import annotations

from enum import Enum

# A type alias (PEP 613 style assignment) — the reader should record it.
Scalar = float


class Kind(Enum):
    """An enum with several members."""

    CIRCLE = "circle"
    RECTANGLE = "rectangle"
    EMPTY = "empty"


class Circle:
    """A class with an initializer, an attribute member, and a method."""

    def __init__(self, radius: Scalar) -> None:
        self.radius = radius

    def area(self) -> Scalar:
        """Instance method calling a module-level function (intra-file call)."""
        return circle_area(self.radius)


def circle_area(radius: Scalar) -> Scalar:
    """Module-level function used by ``Circle.area``."""
    return 3.14159 * radius * radius


class Shape:
    """A base class."""

    def measure(self) -> Scalar:
        return 0.0


class Square(Shape):
    """A subclass — ``class Square(Shape)`` is an ``extends`` edge (Square -> Shape)."""

    def measure(self) -> Scalar:
        return 1.0

File: tests/fixtures/graph-src-tier3/ruby/shapes.rb
# A top-level constant.
PI = 3.14159

# Free function used across files.
def circle_area(radius)
  radius * radius * PI
end

module Geometry
  # A base class.
  class Base
    def kind
      :shape
    end
  end

  # A subclass — Ruby `<` superclass → an `extends` edge (Circle extends Base).
  class Circle < Base
    def area
      3.14
    end
  end

  # A class with methods.
  class Point
    # Constructor.
    def initialize(x, y)
      @x = x
      @y = y
    end

    def distance_to(other)
      hypot(@x - other.x, @y - other.y)
    end

    def self.origin
      Point.new(0, 0)
    end
  end
end

def hypot(a, b)
  Math.sqrt(a * a + b * b)
end
